typing/typecore.ml

```diff
@@ -2602,12 +2642,7 @@ let type_pattern_list
   in
   let patl = List.map2 type_pat spatl expected_tys in
   let pvs = get_ref pattern_variables in
-  let unpacks =
-    List.map (fun { mv_id; mv_name; mv_loc } ->
-      {tu_ident = mv_id; tu_loc = mv_loc; tu_name = mv_name;
-       tu_uid = Uid.mk ~current_unit:(Env.get_unit_name ())}
-    ) (get_ref module_variables)
-  in
+  let unpacks = module_variables_to_unpacks (get_ref module_variables) in
   let new_env = add_pattern_variables !new_env pvs in
```

Comment:
```suggestion
  let unpacks = module_variables_to_unpacks (get_ref module_variables) in
  let new_env = add_pattern_variables !new_env pvs in
  (* Don't call add_module_variables here, because its use of [type_module]
     will fail until after we have type-checked the expressions of the let.
     Example: [let (m : (module S)) = ... in let (module M) = m in ...]
     We learn the signature [S] from the type of [m] in the RHS of the second
     let, and we need that knowledge for [type_module} to succeed.
     If we type-checked expressions before patterns, then we could call
     [add_module_variables] here.
     *)
```

Maybe that's a bit excessive, but I think both you and I thought of adding `add_module_variables` here, and so someone else might wonder why we didn't.

Reply:
I lean on the side of adding a similarly detailed comment, given the opportunity for confusion here. I adopted this comment (almost verbatim). I found it clearer to move the responsibility of calling `add_pattern_variables` also outside of `type_pattern_list`, so I put this comment at the sole callsite of that function.